### PR TITLE
[DevSetting] Replace cinc_version/cinc_installer_url with CliAttributeOverrides which is similar to ExtraChefAttributes

### PR DESCRIPTION
**Pull request by himani2411**
### Description of changes
Replace cinc_version/cinc_installer_url with CliAttributeOverrides

* Model the CLI-side CINC overrides as a single freeform JSON DevSetting (CliAttributeOverrides), mirroring Cookbook.ExtraChefAttributes, instead of two scalar settings. The config exposes cinc_version/cinc_installer_url as read-only accessors that unfurl the blob, so imagebuilder_stack.py and parallelcluster.yaml keep populating CfnParamCincVersion/CfnParamCincInstaller unchanged.

* Add CliAttributeOverridesValidator for JSON well-formedness (object), with a KNOWN_CLI_ATTRIBUTE_OVERRIDES hook for optional future strict key validation.

* Add tests which cover CliAttributeOverridesValidator (valid object, empty/None, malformed JSON, non-object array/scalar) and the ImagebuilderDevSettings cinc_version/ cinc_installer_url accessors that unfurl the JSON blob (full, empty, partial, unknown-keys).

* CincVersion was added https://github.com/aws/aws-parallelcluster/pull/7420

### Tests
* Build Image

### References
* Link to impacted open issues.
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/schemas/imagebuilder_schema.py
Comment (himani2411) [inline cli/src/pcluster/schemas/imagebuilder_schema.py:238]: NOTE FOR REVIEWER: The cli_attribute_overrides can later be moved to BaseDevSettingsSchema when we add more overrides; since the 2 paramters are related to only Build Image I am keeping it in `ImagebuilderDevSettingsSchema`

File: cli/src/pcluster/validators/dev_settings_validators.py [hunk 1]
```diff
@@ -19,6 +19,33 @@
...
+                f"Invalid value in {CLI_ATTRIBUTE_OVERRIDES_PATH}: must be a valid JSON string.",
+                FailureLevel.ERROR,
+            )
+            return
+        if not isinstance(attrs, dict):
+            self._add_failure(
+                f"Invalid value in {CLI_ATTRIBUTE_OVERRIDES_PATH}: must be a JSON object.",
+                FailureLevel.ERROR,
+            )
+        # TODO: optionally validate keys against KNOWN_CLI_ATTRIBUTE_OVERRIDES here.
```
Comment (himani2411) [inline cli/src/pcluster/validators/dev_settings_validators.py]: NOTE FOR REVIEWER: Keeping this as a future improvement since we do not validate all JSON content for ExtraChefAttributes as of today and the cinc attributes that I am replacing also do no have any validators. Any un-related attribute apart from the cinc* one's in the JSON will be a no-op.

File: cli/src/pcluster/validators/dev_settings_validators.py [hunk 2]
```diff
@@ -19,6 +19,33 @@
 ATTR_CLUSTER_READINESS_CHECK_IGNORE_FAILURE = "cluster.cluster_readiness_check_ignore_failure"
 MIN_SLURM_RECONFIGURE_TIMEOUT = 300
 
+CLI_ATTRIBUTE_OVERRIDES_PATH = "DevSettings/CliAttributeOverrides"
+# Keys the CLI currently unfurls from the blob. Kept for an optional future
+# strict check; today validation is well-formedness only (like ExtraChefAttributes).
+KNOWN_CLI_ATTRIBUTE_OVERRIDES = {"cinc_version", "cinc_installer_url"}
```
Comment (gmarciani) [inline cli/src/pcluster/validators/dev_settings_validators.py]: I guess we can remove this constant because we are only interested in enforcing the json formatting of the `CliAttributeOverrides` without making eny assumption or enforcement about its field.

In Cookbook Chef Attribute we introduced the validation of specific attributes only because we decided to vend those attributes publicly to mitigate an impact, but in general we should not do it and keep it free form.

File: cli/src/pcluster/config/imagebuilder_config.py [hunk 3]
```diff
@@ -251,14 +252,36 @@ def __init__(
         self.distribution_configuration = distribution_configuration
         self.terminate_instance_on_failure = Resource.init_param(terminate_instance_on_failure, default=True)
         self.disable_validate_and_test = Resource.init_param(disable_validate_and_test, default=True)
-        self.cinc_installer_url = Resource.init_param(cinc_installer_url, default="")
```
Comment (gmarciani) [inline cli/src/pcluster/config/imagebuilder_config.py:254]: Removing this parameter from DevSettings schema is ok because we do not guarantee backward compatibility for DevSettings by definition. However, I suggest to document the removal in the changelog. the fact that we do not guarantee backward compatibility means it is ok to make this change, but does not mean we should not surface it.

Comment (himani2411) [reply cli/src/pcluster/config/imagebuilder_config.py:254]: IF we never surface the addition of DevSettings in chnagelog; why should we surface their removal?

I propose we do not add this as part of changelog since we never document its existence.

File: cli/src/pcluster/validators/dev_settings_validators.py (hunk 2 above)
Comment (himani2411) [reply cli/src/pcluster/validators/dev_settings_validators.py]: Updated

File: cli/src/pcluster/config/imagebuilder_config.py (hunk 3 above)
Comment (gmarciani) [reply cli/src/pcluster/config/imagebuilder_config.py:254]: The point of a changelog entry is to introduce opportunities and anticipate problems.
Even if we have never advertised the param in our changelog, nothing prevents a user from using it. 
Such user will face an error. Such error cannot be considered a breaking of a contract because it is an unofficial param, but still the entry will allow the user to avoid/motivate the failure.

If you disagree with that I understand because we are on a gray area.
In this case I suggest to let our PM have a call on this.

**Review by gmarciani — APPROVED** [review]
